app/src/main/java/io/appium/settings/utils/GPSTracker.java

```diff
@@ -0,0 +1,109 @@
+/*
+  Copyright 2012-present Appium Committers
+  <p>
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+  <p>
+  http://www.apache.org/licenses/LICENSE-2.0
+  <p>
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+ */
+
+package io.appium.settings.utils;
+
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.location.Location;
+import android.location.LocationListener;
+import android.location.LocationManager;
+import android.os.Bundle;
+import android.os.IBinder;
+import android.util.Log;
+
+public class GPSTracker extends Service implements LocationListener {
+    private static final String TAG = GPSTracker.class.getSimpleName();
+
+    private final Context mContext;
+
+    private static final long MIN_DISTANCE_CHANGE_FOR_UPDATES = 10; // 10 meters
+    private static final long MIN_TIME_BW_UPDATES = 1000 * 60; // 1 minute
+
+    public GPSTracker(Context context) {
+        this.mContext = context;
+    }
+
+    public Location getLocation() {
+        try {
+            final LocationManager locationManager =
+                    (LocationManager) mContext.getSystemService(LOCATION_SERVICE);
+            boolean isGPSEnabled = false;
+            try {
+                isGPSEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
+            } catch (NullPointerException e) {
+                Log.e(TAG, e.getMessage());
+            }
+            boolean isNetworkEnabled = false;
+            try {
+                isNetworkEnabled = locationManager.isProviderEnabled(LocationManager.NETWORK_PROVIDER);
+            } catch (NullPointerException e) {
+                Log.e(TAG, e.getMessage());
+            }
+
+            if (!isGPSEnabled && !isNetworkEnabled) {
+                Log.i(TAG, "Both GPS and Network providers are disabled");
+                return null;
+            }
+
+            Location location = null;
+            if (isGPSEnabled) {
+                locationManager.requestLocationUpdates(LocationManager.GPS_PROVIDER,
```

Comment:
I assume that this implementation is rather unstable - some location updates can be missed.
To catch all updates `requestLocationUpdates` should be called in `onCreate()` method. And the service should handle updates in `onLocationChanged` method.
Like a:
```java
public class GPSTracker extends Service implements LocationListener {
    private static final String TAG = GPSTracker.class.getSimpleName();

    private static Location location = null;
    LocationManager locationManager;
    
    @Override
    public void onCreate() {
        addListenerLocation();
    }
    
    @Override
    public void onDestroy() {
        super.onDestroy();
        locationManager.removeUpdates(this);
    }

    private void addListenerLocation() {
        locationManager = (LocationManager) getSystemService(LOCATION_SERVICE);
        locationManager.requestLocationUpdates(LocationManager.PASSIVE_PROVIDER,
                0, 0, this);
    }
    
    public static Location getLocation() {
        return location;
    }

    @Override
    public void onLocationChanged(Location location) {
        this.location = location;
    }
}
```

Reply:
is it mandatory to register the tracker class in the manifest if we change it this way?

Reply:
Yes, it should be a service like a ```LocationService```.

Reply:
changed